runtime-sdk/src/storage/overlay.rs

```diff
@@ -47,14 +47,17 @@ impl<S: Store> Store for OverlayStore<S> {
         self.parent.get(key)
     }
 
-    fn insert<K: AsRef<[u8]>>(&mut self, key: K, value: &[u8]) {
-        self.overlay
+    fn insert<K: AsRef<[u8]>>(&mut self, key: K, value: &[u8]) -> Option<Vec<u8>> {
+        let last = self.overlay
             .insert(key.as_ref().to_owned(), value.to_owned());
         self.dirty.insert(key.as_ref().to_owned());
+        last
     }
 
-    fn remove<K: AsRef<[u8]>>(&mut self, key: K) {
+    fn remove<K: AsRef<[u8]>>(&mut self, key: K) -> Option<Vec<u8>> {
         // For dirty values, remove from the overlay.
+        // TODO: Has this always been the case, where we diverge from what the Store trait defines
+        // as "remove" in the OverlayStore case?
```

Comment:
I was going to see if it would be reasonable to make Store not throw away the last value in `insert` and `remove`, and I'm distracted by this `remove` implementation in OverlayStore. OverlayStore seems to change the meaning of `remove`. Here `remove` doesn't make future `get` return `None`. Here `remove` sort of resets it to the original value, which might still have been present.

Reply:
No, this does not reset it to the original value as it only removes it from the overlay, but not from the dirty set. So further `get` queries will still return `None`.

Reply:
oh right. None in overlay and key present in dirty means to remove on commit